### PR TITLE
Fix episode wrong prefixes

### PR DESCRIPTION
**Pull request by duehoa1211**
This commit fix for issue #696 on front end
I think we should update the type value of back-end as well to match with front-end EpisodeType

### CODE REVIEW

**Comment by harshithmohan** [comment]
I will let @revam comment

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 1]
```diff
@@ -66,10 +66,36 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
...
+        return '';
+    }
+  };
+
   const airDate = useMemo(() => dayjs(episode.AirDate), [episode.AirDate]);
   const relativeTime = useMemo(() => airDate.calendar(null, CalendarConfig), [airDate]);
   const title = useMemo(
-    () => `${episode.Type === EpisodeTypeEnum.Normal ? '' : episode.Type[0]}${episode.Number} - ${episode.Title}`,
+    () => {
+      if (episode.Type === EpisodeTypeEnum.Normal) {
```
Comment (harshithmohan) [inline src/pages/dashboard/components/EpisodeDetails.tsx:93]: Is this `if` statement required? Because if this doesn't exist the result would be the same anyway. `anidbEpisodeAlias` would return an empty string.

**Comment by harshithmohan** [comment]
We should have prefixes for remaining types also such as PV, Others, Trailers etc

File: src/pages/dashboard/components/EpisodeDetails.tsx (hunk 1 above)
Comment (duehoa1211) [reply src/pages/dashboard/components/EpisodeDetails.tsx:93]: yeah, i will remove that, that is no needed at all

**Comment by duehoa1211** [comment]
Base on [this link](https://wiki.anidb.net/Content:Episodes#Type) classify for each episode type, I don't see the a set prefixes for Unknown, and Interview types. Should I make it as empty string like a normal episode, or I missing something?

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 2]
```diff
@@ -66,10 +66,38 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
...
+    }
+
+    // Prefixes for episode types base on https://wiki.anidb.net/Content:Episodes
+    switch (type) {
+      case EpisodeTypeEnum.OpeningSong:
+        return 'OP';
+      case EpisodeTypeEnum.EndingSong:
+        return 'ED';
+      case EpisodeTypeEnum.Special:
+      case EpisodeTypeEnum.Extra:
```
Comment (revam) [inline src/pages/dashboard/components/EpisodeDetails.tsx:84]: Extra is not used…  yet.

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 3]
```diff
@@ -66,10 +66,38 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
...
+      default:
+        return '';
+    }
+  };
+
   const airDate = useMemo(() => dayjs(episode.AirDate), [episode.AirDate]);
   const relativeTime = useMemo(() => airDate.calendar(null, CalendarConfig), [airDate]);
   const title = useMemo(
-    () => `${episode.Type === EpisodeTypeEnum.Normal ? '' : episode.Type[0]}${episode.Number} - ${episode.Title}`,
+    () => `${anidbEpisodeAlias(episode.Type, episode.Title)}${episode.Number} - ${episode.Title}`,
```
Comment (revam) [inline src/pages/dashboard/components/EpisodeDetails.tsx:100]: If we want to do client-side OP/ED parsing w/variations (e.g. `OP1a`) then I recommend passing in the episode number to the alias function and creating the full string from within the function.

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 4]
```diff
@@ -66,10 +66,38 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
     return `${((resumePosition.asMilliseconds() / duration.asMilliseconds()) * 100).toFixed(2)}%`;
   }, [episode.Duration, episode.ResumePosition]);
 
+  const anidbEpisodeAlias = (type: EpisodeTypeEnum, epTitle: string) => {
```
Comment (revam) [inline src/pages/dashboard/components/EpisodeDetails.tsx:69]: This function seems like it can be moved outside the `EpisodeDetails` component, as it doesn't directly use any of the variables from it's parent scopes.

**Review by revam — CHANGES_REQUESTED** [review]
OPs/EDs parsing is a difficult subject to do correctly, especially when the metadata we get from the xml dumps from AniDB doesn't contain the same definitions they show on their site (e.g. we get Credits 3 (`C3`), but they show Opening 1a (`OP1a`)).

I don't intend to scare @duehoa1211 away, but these are the things I would like changed, and hope it's okay with them to make the changes. :pleading_face:

Also, feel free to provide counter arguments if anyone disagrees with what I wrote.

**Comment by ElementalCrisis** [comment]
Thanks for the PR! 

Let us know if you're having issues making the suggested changes either on here or Discord. :)

**Comment by revam** [comment]
> Base on [this link](https://wiki.anidb.net/Content:Episodes#Type) classify for each episode type, I don't see the a set prefixes for Unknown, and Interview types. Should I make it as empty string like a normal episode, or I missing something?
> 
> Edit: I will update for prefixes has been classified

Forgot to say, but the episode types we have exposed in the v3 api is not a 1:1 from anidb's types. We have more types in case future providers (or shoko itself) will use them (now or) in the future.

**Comment by duehoa1211** [comment]
> OPs/EDs parsing is a difficult subject to do correctly, especially when the metadata we get from the xml dumps from AniDB doesn't contain the same definitions they show on their site (e.g. we get Credits 3 (`C3`), but they show Opening 1a (`OP1a`)).
> 
> I don't intend to scare @duehoa1211 away, but these are the things I would like changed, and hope it's okay with them to make the changes. 🥺
> 
> Also, feel free to provide counter arguments if anyone disagrees with what I wrote.

No worry, i am always welcoming advices to better myself.

> > Base on [this link](https://wiki.anidb.net/Content:Episodes#Type) classify for each episode type, I don't see the a set prefixes for Unknown, and Interview types. Should I make it as empty string like a normal episode, or I missing something?
> > Edit: I will update for prefixes has been classified
> 
> Forgot to say, but the episode types we have exposed in the v3 api is not a 1:1 from anidb's types. We have more types in case future providers (or shoko itself) will use them (now or) in the future.

I noticed we do have `unknown` and `interview`, which are not define in anidb, do you have any ideas should I put for these types?

and when i downloading love live's collection i did see something like live performance as well, do we intent add that as well in the future?

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 5]
```diff
@@ -66,10 +66,38 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
...
+    // TODO: This is a hack, we need update backend to make the EpisodeTypeEnum make more sense.
```
[the remark this answers is not in the document]
Comment (duehoa1211) [reply src/pages/dashboard/components/EpisodeDetails.tsx]: This reply just to document this commit
>we will return 'C' for ThemeSong prefixes as default and if the title has Opening or Ending i will replace with OP or ED and keep the rest to match with AniDB 
>this logic will be same for the future if we gonna use Opening and Ending type since AniDb not gonna return song title anytime soon

File: src/pages/dashboard/components/EpisodeDetails.tsx (hunk 2 above)
Comment (duehoa1211) [reply src/pages/dashboard/components/EpisodeDetails.tsx:81]: I think I just use the prefixes base on AniDb for `Extra` for now

**Comment by revam** [comment]
> I notice we do have `unknown` and `interview`, which are not define in anidb, do you have any ideas should I put for these types?

Don't even map them for now.

> and when i downloading love live's collection i did see something like live performance as well, do we intent add that as well in the future?

Tl;dr: we only support whatever AniDB support for now.

File: src/pages/dashboard/components/EpisodeDetails.tsx [hunk 6]
```diff
@@ -66,10 +66,32 @@ function EpisodeDetails({ episode, isInCollection = false, showDate = false }: P
     return `${((resumePosition.asMilliseconds() / duration.asMilliseconds()) * 100).toFixed(2)}%`;
   }, [episode.Duration, episode.ResumePosition]);
 
+  const anidbEpisodePrefixes = (type: EpisodeTypeEnum, epNumber: number): string => {
```
Comment (revam) [inline src/pages/dashboard/components/EpisodeDetails.tsx:69]: The last change I will request is to move the function out of `EpisodeDetails`, since it doesn't depend on the parent scope at all.

**Review by revam — CHANGES_REQUESTED** [review]

File: src/pages/dashboard/components/EpisodeDetails.tsx (hunk 6 above)
Comment (duehoa1211) [reply src/pages/dashboard/components/EpisodeDetails.tsx:69]: I'm not quite understanding.
Is that you want me to move it out of `EpisodeDetails`'s component, or move it completely out of `EpisodeDetail`'s files, because both of them is count as out of `EpisodeDetails`'s scope?

Comment (revam) [reply src/pages/dashboard/components/EpisodeDetails.tsx:69]: Just out of the component is fine. If you want to know where we place them in the file, then I will recommend looking at other files for optimal placement. The general rule of thumb is to place them above the main component (and any other sub-components inside the file).

From my understanding of the existing component files, then the order of declarations would be;
1. imports,
2. local types/interfaces,
3. local constants,
4. local helper functions, 
5. local helper components,
6. main component, and
7. export main component.

**Review by revam — APPROVED** [review]
